One or two fixes to timeouts

bernt-matthias opened the pull request:
1. do not wait 2x in tool tests
2. do not use accumulated sleep times but time difference 

Not sure if the 2nd is good or possible. If not we should change the planemo CLI docs for the `--test_timeout` parameter.

Edit: 

OK I see `sleep_` is used in testing and also accumulated test times maybe easier to test. So I'm happy to revert https://github.com/galaxyproject/galaxy/pull/14667/commits/cdf7187f583880159aec14da2de59be3c95ade20 .. but I will wait for feedback.

Maybe increasing `delta` a bit will make `--test_timeout` be closer to what the `planemo` user observes.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage 
       **I hope / assume**
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bernt-matthias committed:
only apply timeout once

during tool tests the `wait_for_job` function which waits for a result
or a timeout (given eg by planemo or DEFAULT_TOOL_TEST_WAIT) was applied
twice:

1. in `_verify_outputs`
2. in `verify_output` (which is called by `_verify_outputs`

this made the waiting time twice as large as expected.

I removed the call from the 2nd.

bernt-matthias committed:
make wait_on use real time (not accumulated sleep time)

the function that "realises waiting for a result or timeout"
currently accumulated the sleep times and compared it to the
timeout. this leads to unexpected long test timeouts
(approx 1.5 - 2 times .. independent of the results of the
previous commit: 9006994b46)

Unexpected in the sense of the planemo CLI docs:

```
Maximum runtime of a single test in seconds.
```

Question is if in one of the multiple places where the function
is called accumulated sleep times are really expected (and maybe
we can just change the timeout at these places).

Also removed the sleep_ parameter which is apparently unused
(and would have complicated importing).

jmchilton commented on lib/galaxy/tool_util/verify/wait.py:
```diff
@@ -20,7 +20,6 @@ def wait_on(
     delta: timeout_type = DEFAULT_POLLING_DELTA,
     polling_backoff: timeout_type = DEFAULT_POLLING_BACKOFF,
-    sleep_: Optional[Callable] = None,
 ):
```
this is used by the unit tests to "mock" out sleeping and write unit tests for the parameters. I would keep this configurable.

jmchilton replied:
Although... maybe add a comment since I'll admit it reads a bit funny 😁

bernt-matthias replied:
done

bernt-matthias committed:
restore sleep parameter as _sleep

using the maybe more standard prefix underscore
to mark things that should not be used (also added
comment)

bernt-matthias commented on lib/galaxy/tool_util/verify/wait.py:
```diff
@@ -30,16 +31,18 @@ def wait_on(
+    start = time()
     while True:
+        total_wait = time() - start
+        print(f"{total_wait=} > {timeout=} {total_wait > timeout}")
```
```suggestion
```

bernt-matthias committed:
Remove log messages

mvdbeek commented on test/unit/tool_util/test_wait.py:
```diff
@@ -9,6 +11,7 @@ def __init__(self):
         self.sleeps = []
 
     def sleep(self, delta):
+        sleep(delta)
```
I don't think we need to actually do time.sleep here

bernt-matthias replied:
What would be the alternative?

bernt-matthias replied:
This seems to be last open point here. Could you give me a hint on how to avoid `time.sleep`?

jmchilton replied:
The previous implementation of this method was much more testable. I don't maybe understand why you changed the logic. I guess now you'd need to send in an override of sleep and time and that probably isn't worth it?

bernt-matthias replied:
> I don't maybe understand why you changed the logic.

I noted that the time spent for tests that timeout (controlled via planemo's `--test_timeout` parameter) is a multiple of the time that is specified. If I remember correctly it was off by at least a factor of 2. The reason is that the time needed to execute `function` (repeatedly) is not count in `test_timeout`. 

I found this slightly  annoying. But I completely agree that this makes it harder/impossible(?) to test. 

An alternative might be a reformulation of the help text of planemo's parameter.

If you like you can check the effect with the following "program"

```
import time

class TimeoutAssertionError(AssertionError):
    """Derivative of AssertionError indicating wait_on exceeded max time."""

    def __init__(self, message):
        super().__init__(message)


def wait_on(
    function,
    desc,
    timeout,
    delta=0.25,
    polling_backoff = 0,
    sleep_ = None,
):
    """Wait for function to return non-None value.
    Grow the polling interval (initially ``delta`` defaulting to 0.25 seconds)
    incrementally by the supplied ``polling_backoff`` (defaulting to 0).
    Throw a TimeoutAssertionError if the supplied timeout is reached without
    supplied function ever returning a non-None value.
    """
    sleep = sleep_ or time.sleep
    total_wait = 0.0
    start = time.time()
    while True:
        print(f"{time.time() - start} {total_wait}")
        if total_wait > timeout:
            raise TimeoutAssertionError(TIMEOUT_MESSAGE_TEMPLATE.format(total_wait, desc))
        value = function()
        if value is not None:
            return value
        total_wait += delta
        sleep(delta)
        delta += polling_backoff


def work():
    time.sleep(0.1)

wait_on(work, "blah", 900)
```

mvdbeek commented on lib/galaxy/tool_util/verify/wait.py:
```diff
@@ -20,7 +22,7 @@ def wait_on(
     delta: timeout_type = DEFAULT_POLLING_DELTA,
     polling_backoff: timeout_type = DEFAULT_POLLING_BACKOFF,
-    sleep_: Optional[Callable] = None,
+    _sleep: Optional[Callable] = None,  # only used in tests
```
```suggestion
    sleep_: Optional[Callable] = None,  # only used in tests
```
This being a public function in a published package I wouldn't change the name

bernt-matthias replied:
done

bernt-matthias committed:
rename back sleep param

and add comment

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -210,7 +210,6 @@ def verify_output(self, history_id, jobs, output_data, output_testdef, tool_id,
         attributes = output_testdef.attributes
         name = output_testdef.name
         expected_count = attributes.get("count")
-        self.wait_for_jobs(history_id, jobs, maxseconds)
```
How would this have any impact on the wait time ? The first timeout would raise an exception, so you never get here ?
